SLHCUpgradeSimulations/Geometry/test/Phase2PixelNtuple.cc

```diff
@@ -0,0 +1,908 @@
+		   const int num_simhit,
+		   std::vector<PSimHit>::const_iterator closest_simhit,
+		   const GeomDet* PixGeom);
+  void fillPRecHit(const int detid_db, const int subid,
+		   const int layer_num,const int ladder_num,const int module_num,
+		   const int disk_num,const int blade_num,const int panel_num,const int side_num,
+		   trackingRecHit_iterator pixeliter,
+		   const int num_simhit,
+		   std::vector<PSimHit>::const_iterator closest_simhit,
+		   const GeomDet* PixGeom);
+  std::pair<float, float> computeAnglesFromDetPosition(const SiPixelCluster & cl, 
+						       const PixelTopology  & top, 
+						       const GeomDetUnit    & det ) const; 
+
+private:
+  edm::ParameterSet const conf_;
+  TrackerHitAssociator::Config trackerHitAssociatorConfig_;
+  edm::EDGetTokenT<edmNew::DetSetVector<SiPixelRecHit>> pixelRecHits_token;
+  //edm::InputTag src_;
+  edm::EDGetTokenT<edm::View<reco::Track> > token_recoTracks;
+  bool verbose_;
+  bool picky_;
+  static const int DGPERCLMAX = 100;  
+     
+  //--- Structures for ntupling:
+  struct evt
+  {
+    int run;
+    int evtnum;
+     
+    void init();
+  } evt_;
+   
+  struct RecHit 
+  {
+    int pdgid;
+    int process;
+    float q;
+    float x;
+    float y;
+    float xx;
+    float xy;
+    float yy;
+    float row;
+    float col;
+    float gx;
+    float gy;
+    float gz;
+    int subid,module;
+    int layer,ladder;           // BPix
+    int disk,blade,panel,side;  // FPix
+    int nsimhit;
+    int spreadx,spready;
+    float hx, hy;
+    float hrow, hcol;
+    float tx, ty, tz;
+    float theta, phi;
+  
+    // detector topology
+    int nRowsInDet;
+    int nColsInDet;
+    float pitchx;
+    float pitchy;
+    float thickness;
+
+    // local angles from det position
+    float cotAlphaFromDet, cotBetaFromDet;    
+
+    // digis
+    int fDgN;  
+    int fDgRow[DGPERCLMAX], fDgCol[DGPERCLMAX];
+    int fDgDetId[DGPERCLMAX];
+    //   int fDgRoc[DGPERCLMAX], fDgRocR[DGPERCLMAX], fDgRocC[DGPERCLMAX];
+    float fDgAdc[DGPERCLMAX], fDgCharge[DGPERCLMAX];
+
+    void init();
+  } recHit_;
+
+  edm::Service<TFileService> tFileService; 
+  TTree * pixeltree_;
+  TTree * pixeltreeOnTrack_;
+};
+ 
+ 
+Phase2PixelNtuple::Phase2PixelNtuple(edm::ParameterSet const& conf) : 
+  trackerHitAssociatorConfig_(conf, consumesCollector()),
+  pixelRecHits_token(consumes<edmNew::DetSetVector<SiPixelRecHit>>(edm::InputTag("siPixelRecHits"))),
+  //src_( conf.getParameter<edm::InputTag>( "src" ) ),
+  token_recoTracks(consumes<edm::View<reco::Track> >(conf.getParameter<edm::InputTag>( "trackProducer" ))),  
+  verbose_( conf.getUntrackedParameter<bool>("verbose",false)),
+  picky_(conf.getUntrackedParameter<bool>("picky",false)),
+  pixeltree_(0),
+  pixeltreeOnTrack_(0)
+{
+}
+ 
+ 
+Phase2PixelNtuple::~Phase2PixelNtuple() { }  
+ 
+void Phase2PixelNtuple::endJob() 
+{
+  std::cout << " Phase2PixelNtuple::endJob" << std::endl;
```

Comment:
consider which (if any) of the active cout statements are really necessary, since this will be running as part of a unit test